### PR TITLE
[onnx-to-tosa] add shapehelper for maxpool to support ceil_mode padding

### PR DESCRIPTION
**Pull request by maxbartel**
Maxpool didn't pass the execution test when `ceil = 1`. This adds values to the padding, which should have the same effect.

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 1]
```diff
@@ -25,6 +25,44 @@ namespace onnx_mlir {
...
+  SmallVector<int64_t> kernelShapeVec = extractFromI64ArrayAttr(kernelShape);
+
+  // Stride and pad are optional attributes. If they are non-existing, we get
+  // the standard values.
+  SmallVector<int64_t> stridesVec;
+  if (strides) {
+    stridesVec = extractFromI64ArrayAttr(strides);
+  } else {
+    stridesVec = {1, 1};
+  }
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:46]: You can set the value of stridesVec to {1,1} by default and only modify it if `strides` is set (thus getting rid of the else clause)

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 2]
```diff
@@ -25,6 +25,44 @@ namespace onnx_mlir {
...
+  SmallVector<int64_t, 4> padsVector;
+  if (pads) {
+    padsVector = extractFromI64ArrayAttr(pads);
+  } else {
+    padsVector = {0, 0, 0, 0};
+  }
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:52]: Same thing here

**Review by flemairen6 — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 3]
```diff
@@ -25,6 +25,44 @@ namespace onnx_mlir {
 
 namespace {
 
+// This calculates the values that need to be added to the padding in order to
+// simulate the ceil mode
+llvm::SmallVector<int64_t> getCeilConstants(int64_t ceil,
+    llvm::ArrayRef<int64_t> inputShape, mlir::ArrayAttr kernelShape,
+    mlir::ArrayAttr strides, mlir::ArrayAttr pads) {
+  // This avoids having more if statements when creating the padding const.
+  if (!ceil) {
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:34]: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by ljfitz — APPROVED** [review]
Again, I was caught discussing comments with you that I still had as pending. I'm okqay with this going in, but expect we'll need to adjust to the shape support before we have the conversion code that is acceptable upstream.

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 4]
```diff
@@ -53,12 +89,9 @@ class ONNXMaxPoolSingleOutOpLoweringToTOSA
...
     }
     auto InputType = Input.getType().cast<TensorType>();
-    auto resultType = op.getResult().getType();
-    auto resultShape = resultType.cast<TensorType>().getShape();
     // Construct the transposed type for the new MaxPool OP
-    Type newResultType = RankedTensorType::get(
-        {resultShape[0], resultShape[2], resultShape[3], resultShape[1]},
-        InputType.getElementType());
+    Type newResultType =
+        RankedTensorType::get({-1, -1, -1, -1}, InputType.getElementType());
```
[the remark this answers is not in the document]
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:90]: We had this dicussion here https://github.com/ljfitz/onnx-mlir/pull/74#discussion_r1004482549 and Ferdinand added warnings for them in line 97 and 101. I can see them if I click the blue bar below this box

[the remark this answers is not in the document]
Comment (ljfitz) [reply src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:90]: We should have used `rewriter.notifyMatchFailure` there, and anywhere else we meet an attribute that would have effect on the conversion but we are not yet willing to implement.

**Comment by maxbartel** [comment]
This is now quite a big refactoring bringing in new functionality. I would appreciate feedback on this :)

**Review by flemairen6 — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/DialectBuilder.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ */
+
+//====------ DialectBuilder.hpp - TOSA dialect builder --------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file contains the dialect build for the TOSA dialect. Uses the same
+// implementation as MHLO with minor differences.
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/DialectBuilder.cpp:13]:
```suggestion
// implementation as ONNXToMhlo with minor differences.
```

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp [hunk 6]
```diff
@@ -27,14 +27,23 @@
...
 
-Value sliceTensor(PatternRewriter &rewriter, Operation *op,
-    Value &inputConst, const llvm::ArrayRef<int64_t> &size,
-    const llvm::ArrayRef<int64_t> &start) {
+llvm::SmallVector<int64_t> createInt64VectorFromIndexExpr(
+    llvm::ArrayRef<IndexExpr> indexVector) {
+  llvm::SmallVector<int64_t, 4> literalVector;
+  for (const auto &indexExpr : indexVector) {
+    literalVector.push_back(indexExpr.getLiteral());
+  }
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.cpp:41]: I'm not sure I got the syntax right, but the use of `llvm::transform` is preferable to such loops.
```suggestion
llvm::transform(indexVector, literalVector.begin(), [](const auto &indexExpr) { indexExpr.getLiteral(); });
```

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 7]
```diff
@@ -15,95 +15,138 @@
...
+  SmallVector<int64_t, 2> kernelShapeVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.kernelShape);
+
+  // Get stride and pad vectors.
+  SmallVector<int64_t, 2> stridesVec = shapeHelper.strides;
+  SmallVector<int64_t, 4> padsVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // Check if the idiv_check for the output dimentions in
+  // https://www.mlplatform.org/tosa/tosa_spec.html#_max_pool2d has no rest. If
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:48]: Eek - I don't know what `has no rest` means. Is this a defined term?

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 8]
```diff
@@ -15,95 +15,138 @@
...
-    if (autoPad != "NOTSET") {
-      op.emitWarning(
-          "auto_pad attribute is deprecated and its value will be ignored");
-    }
     if (dilations) {
-      op.emitWarning("dilations attribute is unsupported by TOSA and its value "
-                     "will be ignored");
+      return rewriter.notifyMatchFailure(maxpoolOp,
+          "dilations attribute is unsupported by TOSA and its value "
+          "will be ignored");
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:109]:
```suggestion
          "dilations attribute is unsupported by TOSA");
```

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 9]
```diff
@@ -15,95 +15,138 @@
...
           op, "memrefs as inputs are unsupported by TOSA");
     }
-    auto InputType = Input.getType().cast<TensorType>();
-    auto resultType = op.getResult().getType();
-    auto resultShape = resultType.cast<TensorType>().getShape();
+    auto inputType = input.getType().cast<TensorType>();
+
+    if (inputType.getShape().size() != 4) {
+      return rewriter.notifyMatchFailure(
+          maxpoolOp, "TOSA only supports tensors with size 4");
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:98]:
```suggestion
          maxpoolOp, "TOSA only supports maxpool 2d");
```

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp [hunk 10]
```diff
@@ -15,95 +15,138 @@
...
+      return rewriter.notifyMatchFailure(maxpoolOp,
+          "dilations attribute is unsupported by TOSA and its value "
+          "will be ignored");
     }
     if (storageOrder && storageOrder.getSInt() != 0) {
-      op.emitWarning("storage_order attribute is unsupported by TOSA and its "
-                     "value will be ignored");
+      return rewriter.notifyMatchFailure(maxpoolOp,
+          "storage_order attribute is unsupported by TOSA and its "
+          "value will be ignored");
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:114]:
```suggestion
          "storage_order attribute is unsupported by TOSA");
```

**Review by ljfitz — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp (hunk 7 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:48]: Oh that was a 1:1 german to english translation that is not possible. The correct term is remainder

**Pull request merged**